### DISCUSSION TITLE
Looking for a way to track if the portal is up or not

### DISCUSSION THREAD

**Discussion by softwaregravy** [Hieromon/AutoConnect#543] [Feature Request]
I'm looking for a way to determine if the portal is available or not from code. We'd like to display an icon/change visible behavior if the portal is available.

I've hooked in to the `onDetect` callback for starting the portal, and then we have the `onConnect`. With a successful connection, my code works as I put a `portal_available = true` in the first callback, and a `portal_available = false` in the second. However, a `_stopPortal()` call happens for a few other reasons like timeout/disconnect or reset. I'm not seeing any way to get a hook to know we shut down. 

One avenue I've looked into is the WifiClient on the WebServer, but this only works if someone has actually connected. I'm looking for if the portal is available at all.
```c++
// assume we have this
static WebServer   server;
static AutoConnect auto_connect(server);

// and we are trying to find whether the portal is up
WifiClient wifi_client = server.client();
// I assume this is true only if someone has connected to the portal. Maybe I'm wrong?
wifi_client.connected(); 
```
[Webserver.client()](https://github.com/espressif/arduino-esp32/blob/master/libraries/WebServer/src/WebServer.h) 
[WifiClient.connected()](https://github.com/espressif/arduino-esp32/blob/bd71449988fbb7dd55e533ced895b166a41c4736/libraries/WiFi/src/WiFiClient.h)

I also looked into  [`WebServer`](https://github.com/espressif/arduino-esp32/blob/master/libraries/WebServer/src/WebServer.h#L177) which has a [`WifiServer`](https://github.com/espressif/arduino-esp32/blob/bd71449988fbb7dd55e533ced895b166a41c4736/libraries/WiFi/src/WiFiServer.h) underneath it which does have an `operator bool()` which I think would give me what I'm looking for. But I don't see a good way to get to that object from `AutoConnect` or `WebServer`. `WebServer` doesn't seem to expose this instance at all.

Thoughts?

An alternative would be to ask for another callback on AutoConnect probably from inside `_stopPortal()`, or for AutoConnect itself to keep track of the status and provide a bool method to check. 

Thoughts on this?

Thanks

**Comment by Hieromon** [comment]
@softwaregravy Your question may lead to an AutoConnect enhancement. I would like to move this topic to Discussions and discuss specific implementations there.
Thank you.

**Comment by softwaregravy** [comment]
I think the easiest implementation might be for AutoConnect to keep track of the state itself. With a simple boolean. The question is where to set this. 

I might suggest setting to true in `_softAP` and then to false in `_stopPortal` and `handleRequest`
* for `_softAP`. I think this covers both the `begin` and `handleRequst` use cases, which I think are the only 2 places the portal starts
* the `_stopPortal` case covers stopping after unsuccessful connections, and then also in `handleRequest` covers successful connections on [line 592](https://github.com/Hieromon/AutoConnect/blob/master/src/AutoConnectCoreImpl.hpp#L592)

Then we just expose this via a simple boolean check `isPortalAvailable` or something.

**Comment by softwaregravy** [comment]
For illustrative/discussion purposes: #545 

I found that the cleanest place to integrate the checks was in the DNS control. I'm testing this out in my project.

**Comment by Hieromon** [comment]
@softwaregravy Thanks for your preliminary survey and PR.
Before designing a new API, I have traced your mention below. This is to get an accurate picture of your situation and what you want. If my understanding differs from your goals, please point it out.

> a _stopPortal() call happens for a few other reasons like timeout/disconnect or reset. I'm not seeing any way to get a hook to know we shut down.

Yes, you are correct. Captive Portal - i.e., just a spoofed response to a DNS lookup for Internet connection verification that occurs on a new connection attempt from the client device; it needs a DNS server and SoftAP to work, so if we can detect that one or the other is missing, I think we can fulfill your request. So, about your approach as follows:

> I think the easiest implementation might be for AutoConnect to keep track of the state itself. With a simple boolean.
> I might suggest setting to true in _softAP and then to false in _stopPortal and handleRequest

I believe your PR #545 meets the above requirements but would suggest adding further detection conditions. It is an approach that allows the _isPortalAvailable_ function to determine state at a point in time, rather than a DNS and SoftAP state trace on AutoConnect itself.

In _AutoConnectCore\<T\>::isPortalAvailable_ without `_portalAvailable` flag:
 
- Retrieval a logical value of `WiFi.getMode() & WIFI_MODE_AP`
- `WiFi.softAPIP()` is identical to `_apConfig.apip`
- `_dnsServer` is not `nullptr`

`isPortalAvailable()` is `true` when the above conditions are met. This is semantically equivalent to your approach. And it can keep track of the captive portal availability even when SoftAP is intentionally deactivated outside of AutoConnect.
The pseudo-code for _isPortalAvailable_ would probably is following:

```cpp
bool isPortalAvailable(void) {
  bool state = WiFi.getMode() & WIFI_MODE_AP;
  state &= (WiFi.softAPIP() == _apConfig.apip);
  state &= (_dnsServer.get() != nullptr);
  return state;
}
```

How about this?

**Reply by softwaregravy** [reply]
I think getting to a solution without adding another variable is a much better approach. 👍 

I'm probably not as familiar with the call to `softAPIP()`. I'm sure you're much more familiar. For my reading, it [seems](https://github.com/espressif/arduino-esp32/blob/ba6e82c30d31a2ddc1276a045e0c42a29b38869f/libraries/WiFi/src/WiFiAP.cpp#L258) to be a thin wrapper around `esp_netif_get_ip_info` which [seems](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/network/esp_netif.html) like it may not be indicative of current state:

```c++
IPAddress WiFiAPClass::softAPIP()
{
    if(WiFiGenericClass::getMode() == WIFI_MODE_NULL){
        return IPAddress();
    }
	esp_netif_ip_info_t ip;
    if(esp_netif_get_ip_info(get_esp_interface_netif(ESP_IF_WIFI_AP), &ip) != ESP_OK){
    	log_e("Netif Get IP Failed!");
    	return IPAddress();
    }
    return IPAddress(ip.ip.addr);
}
```

Docs for `esp_netif_get_ip_info`
> If the interface is up, IP information is read directly from the TCP/IP stack. If the interface is down, IP information is read from a copy kept in the ESP-NETIF instance

If the portal is brought up, shut down, and then brought up again, is there any value in the check to the `_apConfig.apip`?

**Comment by Hieromon** [comment]
@softwaregravy 
> it seems to be a thin wrapper around esp_netif_get_ip_info which seems like it may not be indicative of current state:
> If the portal is brought up, shut down, and then brought up again, is there any value in the check to the _apConfig.apip?

I understand your concern. The [`esp_netifs static structure`](https://github.com/espressif/arduino-esp32/blob/972c3bbd00e7294739ba9b5eea0db3a0298d0b53/libraries/WiFi/src/WiFiGeneric.cpp#L56
) is only initialized when [`tcpInit()`](https://github.com/espressif/arduino-esp32/blob/972c3bbd00e7294739ba9b5eea0db3a0298d0b53/libraries/WiFi/src/WiFiGeneric.cpp#L610) is run from the [`wifiLowLevelInit()`](https://github.com/espressif/arduino-esp32/blob/972c3bbd00e7294739ba9b5eea0db3a0298d0b53/libraries/WiFi/src/WiFiGeneric.cpp#L653). Usually, this is a WiFi module activation within ESP32 and is also performed by WIFI_MODE_STA only. However, [`set_esp_interface_ip()`](https://github.com/espressif/arduino-esp32/blob/972c3bbd00e7294739ba9b5eea0db3a0298d0b53/libraries/WiFi/src/WiFiGeneric.cpp#L88) overwrites it with the new IP given in the argument, so `WiFi.softAPIP()` and `_apConfig.apip` will be in sync if [`WiFiAPClass::softAPConfig()`](https://github.com/espressif/arduino-esp32/blob/972c3bbd00e7294739ba9b5eea0db3a0298d0b53/libraries/WiFi/src/WiFiAP.cpp#L207) is executed reliably in _AutoConnectCore\<T\>::\_softAP()_.

On other hand, stopping SoftAP using `WiFi.mode(~WIFI_MODE_AP)` only will probably cause `softAPIP()` still return the equivalent `_apConfig.apip`. In that case, if the SoftAP is started outside of AutoConnect and is configured with the same IP as the ESP32AP, it will seem like the portal is open. But the DNSServer remains shut down, so Internet transparency validation requests from client devices pass through. This use case assumes the coexistence of ESP-NOW and AutoConnect. Related to #210.

So, I think it is probably okay.

**Comment by softwaregravy** [comment]
Seems solid. 👍

**Comment by Hieromon** [comment]
@softwaregravy 
For now, I have implemented the `bool AutoConnectCore<T>::isPortalAvailable()`. You can evaluate it in the [**patches/543**](https://github.com/Hieromon/AutoConnect/tree/patches/543) branch. Please give it a try. Let me know if there are some  problems in your use case.
I plan to release AutoConnect v1.4.1 soon, could you wait until v.1.4.1 release for this enhancement?

Oh, yes. A little bonus: If you build by giving the ~~`AC_USE_PORTALIDENTIFER`~~ `AC_SHOW_PORTALIDENTIFER` macro in `AutoConnectDefs.h`, you will see the following checkmark in the captive portal state. In PlatformIO, it is enabled by specifying ~~`build_flags=-DAC_USE_PORTALIDENTIFIER`~~ `build_flags=-DAC_SHOW_PORTALIDENTIFIER`.


![ac](https://user-images.githubusercontent.com/12591771/207793942-e8acbb25-f1da-4ad3-ba8d-d56f7252814e.png)

![el](https://user-images.githubusercontent.com/12591771/207794097-c37fccca-4f0e-4f56-a5ee-f4d46d0d2e3b.png)

**Comment by softwaregravy** [comment]
The boolean check appears to be working well. 👍 

The check-mark has limited utility for me. We don't connect to the weberver unless the portal is active, so it's just always checked.

Thank you!

PS. I tested on a local implementation by copying your suggestion. Didn't actually test your patch branch.

**Reply by Hieromon** [reply]
@softwaregravy, Glad it worked out for your use case. Indeed, the checkmark is just an indicator, so its usefulness will be limited.
I will release it as AutoConnect v1.4.1 after validation with my test cases.

Thank you for your contribution.

**Comment by Hieromon** [comment]
This feature was released in AutoConnect [v1.4.1](https://hieromon.github.io/AutoConnect/api.html#isportalavailable).
